src/foam/nanos/notification/push/APNSPushService.js

```diff
@@ -0,0 +1,165 @@
+          payloadBuilder.setAlertTitle((String) msg.get("title"));
+          String payload = payloadBuilder.build();
+          // Add a ttl for notifications on the payload
+          String token = TokenUtil.sanitizeTokenString(sub.getEndpoint());
+          
+          // TTL for notification delivery, after this time apns will stop trying to deliver this notification
+          // Currently hardcoded to 7 days 
+          Instant instant = Instant.now(); 
+          instant = instant.plus(7, ChronoUnit.DAYS); 
+          SimpleApnsPushNotification pushNotification = new SimpleApnsPushNotification(token, cred.getAppBundleId(), payload, instant);
+
+         try {
+            // Asking the APNs client to send the notification 
+            // and creating the future that will return the status 
+            // of the push after it's sent.  
+            final PushNotificationFuture<SimpleApnsPushNotification, PushNotificationResponse<SimpleApnsPushNotification>> sendNotificationFuture = ((com.eatthepath.pushy.apns.ApnsClient) getApnsClient()).sendNotification(pushNotification);  
+
+            // getting the response from APNs 
+            final PushNotificationResponse<SimpleApnsPushNotification> pushNotificationResponse = sendNotificationFuture.get();
+            if (pushNotificationResponse.isAccepted()) {
+                // TODO: Replace with log
+                System.out.println("Push notification accepted by APNs gateway.");
+            } else {
+                //TODO: Replace with alarms
+                System.out.println("Notification rejected by the APNs gateway: " +
+                        pushNotificationResponse.getRejectionReason());
+
+                if (pushNotificationResponse.getTokenInvalidationTimestamp() != null) {
+                    System.out.println("\t…and the token is invalid as of " +
+                        pushNotificationResponse.getTokenInvalidationTimestamp());
+                }
+            }
+          } catch (final ExecutionException e) {
+           // Should retry if it fails without explanation i.e. we never get a response from apple's servers however a rejection from the apple server should be considered permanent and no retry should be attempted
+            System.err.println("Failed to send push notification.");
+            e.printStackTrace();
+            send(sub, msg);
+          } catch (final Exception e) {
+            // TODO: Replace with loggers
+            System.err.println("Failed to send push notification.");
+            e.printStackTrace();
+          }
+      `
+    },
+
+    {
+      name: 'getCredentials',
+      args: 'X x, String spid',
+      type: 'APNSCredential',
+      javaCode: `
+        APNSCredential credentials = null;
+        DAO credentialDAO = (DAO) x.get("credentialsDAO");
+        ArraySink arraySink = new ArraySink();
+        credentialDAO.where(MLang.EQ(APNSCredential.SPID, spid)).select(arraySink);
```

Comment:
this is not sufficient ... there may over time be multiple different credential models in credential dao for any given spid
better to use the service name

Reply:
Okay, should I add the prod credential too? or will you do that?